### ISSUE TITLE
Change IDE to other screen > task symbol stays on the fist screen (same for built applications)

### ISSUE THREAD

**Issue by ElHado** [twinbasic/twinbasic#1765]
**Describe the bug**
I have 3 screens (one of the laptop, two attached via HDMI and DP). If I change other programs then my taskbar updates the icon to the correct screen. But with TB IDE and applications this don't work. It stays always on the same (first screen) taksbar.
**To Reproduce**
1. Open TB
2. move it to an other screen
3. check if icon change to the new screen or not

**Expected behavior**
TB-icon should show up on the correct screen

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro
 - twinBASIC compiler version 0.15.437

**Comment by ElHado** [comment]
At least for built applications this should work I think. Any ideas for a workaround till this is fixed?

**Comment by fafalone** [comment]
VB6 shows up on the correct screen?

**Comment by ElHado** [comment]
I don't have VB6 running anymore so I can't check with IDE. But a application built in VB6 shows the same behavior: It shows up only at the main taskbar. Never noticed this. 🤔

But I think it should be like the most other applications and show where the application is. (or?)

**Comment by fafalone** [comment]
That would fall under feature request, rather than bug then.

Looking into this more; I think the issue is your main app window is owned not by the desktop but by ThunderMain, a hidden window that VB, and presumably tB, use to implement some features.

I don't know what, if any, the consequences would be for changing that.

**Comment by ElHado** [comment]
Yes, you're right. In this case it would be a feature request. 

Maybe this is possible to implement at some point in the future. 👍

**Comment by fafalone** [comment]
I asked Wayne, and it's doable, but it would have to be opt-in, as it might break compatibility with existing VB6 code relying on the window order/ownership. 

Perhaps we could get a Form option, 'OwnedByDesktop'?

**Comment by ElHado** [comment]
Thank you fafalone! Would be great if we could get this. I think this would be the expected behavior for the users of our applications. 

And if we can change this behavior we can go back if it breaks something.

**Comment by fafalone** [comment]
I made a formal feature request for this. In the meantime, try seeing if changing it during runtime works:

```vba
Private Enum GWL_INDEX
    GWL_WNDPROC = (-4)
    GWL_HINSTANCE = (-6)
    GWL_HWNDPARENT = (-8)
    GWL_ID = (-12)
    GWL_STYLE = (-16)
    GWL_EXSTYLE = (-20)
    GWL_USERDATA = (-21)
End Enum

#If Win64 Then
Private DeclareWide PtrSafe Function SetWindowLong Lib "user32" Alias "SetWindowLongPtrW" (ByVal hWnd As LongPtr, ByVal nIndex As GWL_INDEX, ByVal dwNewLong As LongPtr) As LongPtr
#Else
Private DeclareWide Function SetWindowLong Lib "user32" Alias "SetWindowLongW" (ByVal hwnd As Long, ByVal nIndex As GWL_INDEX, ByVal dwNewLong As Long) As Long
#End If
Private DeclareWide PtrSafe Function GetDesktopWindow Lib "user32" () As LongPtr

    Private Sub Form_Load() Handles Form.Load
        SetWindowLong Me.hWnd, GWL_HWNDPARENT, GetDesktopWindow()
    End Sub
```

**Comment by EduardoVB** [comment]
PS: I've been researching how to handle the issue and found nothing. I wanted to "fix" that behavior for my VB6 programs, but nothing that I tried worked.

Also, you can see that there is the issue of the taskbar icon size. tB shows a too small icon in my 300% DPI TV (my second monitor). VB6 even smaller.

**Comment by fafalone** [comment]
From OPs question he's talking about the same thing you are: Other apps move to the new monitor when that option is enabled. But VB6 and tB apps do not. 

To be clear, Eduardo, you're saying there is, or *is not*, a difference between VB/tB and other apps in this behavior?

I read it as there is a difference, and in researching this issue, hidden windows that actually owned the main window instead of the desktop were the cause.

I don't have 2 monitors to test.

**Comment by fafalone** [comment]
Setting it runtime isn't helping changing the outcome; all the answers I found though did set it during window creation. But I was able to confirm there is a tB/VB6 difference between those and other apps; so if not ThunderMain, then what?

**Comment by EduardoVB** [comment]
tB and VB6 behave exactly the same: the forms taskbar icons (for each form individually) show up in the taskbar of the monitor where they first appear, and do not move to another monitor when the form is dragged to that monitor.

**Comment by EduardoVB** [comment]
In the example that I posted you can see that there are two forms: one shown in the primary monitor and the other one in a secondary monitor. And their taskbar icons appear one on each monitor (and they can be different icons -at least in VB6, that I tested that-, it is set in the Icon property of each form).
So, that's an evidence that the issue can't be related to the ThunderMain window.

**Comment by fafalone** [comment]
I'm not understanding where you're getting this can't be related to ThunderMain. Both do the same thing with a hidden window owner instead of the desktop, both act the same, but you're saying it can't be because of the thing they have in common everyone says causes this? It wouldn't matter VB's is named ThunderRT6Main instead.

**Comment by EduardoVB** [comment]
What I'm saying is that the taskbar icon does not come from the ThunderMain window (but from the form itself).
If you are saying something else, then it is me that I don't understand.

**Comment by fafalone** [comment]
Yes we're talking about whether the taskbar icon stays on its initial monitor, or follows the app to another monitor when the option to do that is enabled.

Both VB6 and tB do not. Delphi apps by default do not. Most other apps do. The thing the three of these have in common, the form whose icon is on the taskbar, it is owned by a hidden window (VB6: ThunderRT6Main, tB: ThunderMain), rather than by the desktop like most apps.

I've now confirmed this issue, but I'm CONFUSED AS HELL:

Everything I read says `HWND_DESKTOP` (0) is only correct for a few APIs that specifically document that as what they use to represent it. And that you should use `GetDesktopWindow()` for all other purposes when you want the desktop window. But I've now gotten tB taskbar icons to follow the window from monitor to monitor, by setting the owner to 0. So I guess 0 *isn't* interpreted as the desktop hwnd, and 0 and GetDesktopWindow aren't equivalent owners? So then if  tB provides this option, it shouldn't be called OwnedByDesktop I guess???


-----
WORKAROUND\
This now works to get the icon to follow:

```vba
Private Enum GWL_INDEX
    GWL_WNDPROC = (-4)
    GWL_HINSTANCE = (-6)
    GWL_HWNDPARENT = (-8)
    GWL_ID = (-12)
    GWL_STYLE = (-16)
    GWL_EXSTYLE = (-20)
    GWL_USERDATA = (-21)
End Enum

#If Win64 Then
Private DeclareWide PtrSafe Function SetWindowLong Lib "user32" Alias "SetWindowLongPtrW" (ByVal hWnd As LongPtr, ByVal nIndex As GWL_INDEX, ByVal dwNewLong As LongPtr) As LongPtr
#Else
Private DeclareWide Function SetWindowLong Lib "user32" Alias "SetWindowLongW" (ByVal hwnd As Long, ByVal nIndex As GWL_INDEX, ByVal dwNewLong As Long) As Long
#End If

    Private Sub Form_Load() Handles Form.Load
        SetWindowLong Me.hWnd, GWL_HWNDPARENT, 0
    End Sub
```

**Comment by EduardoVB** [comment]
Yes, it seems that you are right.

It is weird the effect of not moving the icon (when they have the owner window), since when there are two forms that appear on different monitors, they show each icon on the corresponding monitor.

**Comment by bclothier** [comment]
> CONFUSED AS HELL

FWIW, Microsoft has this in their documentation for [`SetWindow***` functions](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowlongw):

> You must not call SetWindowLong with the GWL_HWNDPARENT index to change the parent of a child window. Instead, use the [SetParent](https://learn.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-setparent) function.

You would use `SetParent` where it explicitly say to use `NULL` to set to desktop:

> A handle to the new parent window. If this parameter is NULL, the desktop window becomes the new parent window.

I would not be surprised if passing `GWL_HWNDPARENT` was actually calling `SetParent` behind the scene. I have not tried calling `SetParent` passing the desktop handle but even if that worked, it would point to the fact that changing parent via the `SetWindowLong***` function is glitchy.

**Comment by EduardoVB** [comment]
SetWindowLong with GWL_HWNDPARENT and SetParent do completely different things.
I think that is the reason of the "warning" about not using SetWindowLong with GWL_HWNDPARENT but SetParent instead.

It was MS that caused that confusion by calling "parent" to two different things.

For the sake of clarity it is better to call "parent" to the relation of a window with its controls, like a form with a command button, and to call "owner" to the relation of a window with another window that stays always in the back of it. Like what you get with `Me.Show, SomeForm`

Then, GWL_HWNDPARENT should have been named GWL_HWNDOWNER, but it is too late to make that change.

**Comment by wqweto** [comment]
> I would not be surprised if passing `GWL_HWNDPARENT` was actually calling `SetParent` behind the scene.

Btw, it's the other way around -- `SetParent` sets `GWL_HWNDPARENT` long **and** redraws whichever hwnd is neccessary to refresh new window heirarchy on screen.

**Comment by EduardoVB** [comment]
https://devblogs.microsoft.com/oldnewthing/20100315-00/?p=14613

**Comment by fafalone** [comment]
SetParent sets the *parent*. GWL_HWNDPARENT is poorly named: it sets the *owner*, which is different from the parent. 

The warning is aimed at not confusing the two. 99% of the time, you want to set the parent. This is the other 1%.
